### ISSUE TITLE
High cpu_system with many open connections

### ISSUE THREAD

**Issue by chbuescher** [grobian/carbon-c-relay#216]
Hi again,
in the meantime - see issue #126 - we have about 6.500 open connections and 2.5 Mio metrics every 15 seconds. The cpu usage increased to 240% with 8 workers (with 16 workers it went over 300%). I looked for a solution to handle even more connections, for we are planning another rollout. I found libevent, a library for handling a large number of open connections. After promising tests of a simple libevent-based TCP server i modified carbon-c-relay to use libevent. Now i have a working version based on version 2.2, that - after flawlessly running in our test env for one week - we now run in production. The cpu usage in production went down to 60% - in the test env (2.900 connections) from 50% to 20%. In my tests the libevent based relay scales linearly up to 15.000 connections when testing with 100 metrics per connection and an interval of 15 seconds.
What you loose is a bit of control: No throttling of fast event sources. And we don't use aggregation, so we haven't tested that. Up to number of connections of about 1000 your solution seems to be a bit more efficient.
Are you interested in my solution ? Diff, python stress script ...

**Comment by grobian** [comment]
It sounds interesting to say the least.  I'd like to see what you did, it sounds like it is an improvement nonetheless!

**Comment by chbuescher** [comment]
[relay-libevent.zip](https://github.com/grobian/carbon-c-relay/files/492733/relay-libevent.zip)
In the zip file you can find the diff to version 2.2/commit 2e76c18c3d7387ee689ee233ec4c65fa68eee303 and the files to stress test the relay: a dummy tcp socket server listening on port 5555, the stresser script, that opens many sockets to the graphite relay and the test configuration for the relay.
Most changes are made in dispatcher.c, the other changes are due to the interface changes in the dispatcher functions.
We still use RHEL6 on our servers, and the libevent library 1.4.x isn't threadsafe, so all the libevent stuff must be done in one thread, the worker 0 : the libevent callback functions dispatch_accept_cb_xxx when there is a new connection, the callback function dispatch_connread_cb for changes in the state of a client connection and the timer callback dispatch_timer_cb for un-/registering the listener sockets.
I have moved the reading of the socket data and preparsing of the read data - searching the last newline char - into dispatch_connread_cb to ensure fast reading of new data and freeing most of the buffer as soon as possible. The new data is then copied to a new struct called metric_block and then inserted in the todo queue of one of the workers.
All connections are handled inside libevent, so i removed the connections list completely - and the udp connections are now handled by the listeners list.
The cpu time in the worker 0 libevent callbacks is measured by the new counter listener_wallTime_us, but i can't measure the libevent internal loop stuff.

**Comment by Civil** [comment]
To overcome issues with older libevent (thread-safeness) you might want to create a configuration system and in case if libevent <2.0 static link carbon-c-relay with bundled version.

**Comment by chbuescher** [comment]
I handle all libevent stuff in only one thread, and i have validated it with helgrind and drd, so i don't expect any thread-safety issues.

**Comment by Civil** [comment]
> I handle all libevent stuff in only one thread, and i have validated it with helgrind and drd, so i don't expect any thread-safety issues.

Yeah, I only meant that some distros (ubuntu LTS, Debian Stable) ships libevent 2.0 and new versions of libevent are thread-safe and you can keep the code multi-threaded.

**Comment by chbuescher** [comment]
Found a segmentation fault when using aggregations: dispatch_initlisteners() must be called before the invocation of router_readconfig() in relay.c

**Comment by Civil** [comment]
I just wanted to say that we ended up rebasing @chbuescher patchset on top of v2.6 (the last one that was stable for us) and it helped to solve one of our performance problem with one of the clusters.

Briefly about the problem - one of the metric producers started sending twice more data in a single connection (we are talking about 30 connections per relay, and bulk sends of 2-3M points every 10 seconds from each box). And then they realized that instead of 2 seconds it now takes 6 minutes to send data out.

With this patch it now takes 4 seconds to accept and relay the data.

**Comment by chbuescher** [comment]
PCRE only helps if you use matches (blackhole, aggregate...). Recently we enabled the PCRE JIT compiler - just by using the option PCRE_STUDY_JIT_COMPILE in pcre_study (in router.c). The worker threads now use 50% less cpu time with our configuration.

**Comment by Civil** [comment]
PCRE was just nice to have, that's it. It haven't helped our case a lot.

Though I think while porting that to 2.6 I've accidentally reverted some fixes for 'send statistics', cause it triggered one of the bugs we had while using 2.3-2.5 releases. But I'll investigate that later.

**Comment by grobian** [comment]
Send statistics has a bug?

**Comment by Civil** [comment]
In the version that @chbuescher patched. For some reason it resurrects one of the old bugs when statistics is malformed.

**Comment by chbuescher** [comment]
I have updated my fork of your project to version 2.6 and enabled PCRE JIT compile.
It works fine in our preproduction environment.
But version 2.6 had a bug with send statistics and blanks ...

**Comment by Civil** [comment]
That bug with malformed sends doesn't happen if I use ```memchr``` instead of ```memrchr```.

**Comment by Civil** [comment]
The way how the conn->buff fills allows it to have incomplete metrics at this point, but only in case if the batch is large.

I'm able to reproduce that by specifying 300 servers (splitted by 10 clusters) in the config, then carbon-c-relay's own stats are larger than buff and metrics got appended one to each other.

**Comment by chbuescher** [comment]
Using memchr will have a  performance impact. The idea is, that worker 0, the thread that does all libevent/socket work, does only minimal parsing of the read data (split it at the last separator char) and puts the data on a queue for consumption by the other worker threads. Using memchr will split the data at each separator and put each each line on the queue.

**Comment by chbuescher** [comment]
The statistics collector sends its metrics through a pipe to the dispatcher. Unfortunately there seem to more then one open connections for this pipe, and therefore several buffers. So the dispatch_connread_cb does not append the new metrics to the right buffer.

**Comment by Civil** [comment]
I think it's writing data to that pipe in one thread, but there are some problems with how you store partly written data.

**Comment by chbuescher** [comment]
I'm pretty sure that partial data handling works ok with tcp connections. I've committed some patches in my fork, so that collector and aggregator send theirs metrics via tcp connection over localhost. Please give it a try.

**Comment by Civil** [comment]
Yeah, sending over TCP seems to fix the problem

**Comment by RudW0lf** [comment]
Hello. I have the same issues with original carbon-c-relay. Forked carbon-c-relay solved my issue. 
Our agent send data one large batch (2-3M per 10-20 secs). Non-patched version dropped huge part of metrics (see attached screenshot).
![image](https://user-images.githubusercontent.com/17278696/36478151-2a1940cc-171d-11e8-8087-4e1bf09bcc15.png)

Will these patches be included in the main branch?

**Comment by grobian** [comment]
They quite diversed from the main branch, that makes it somewhat non-trivial.  Intention, however, is to include them or the idea behind it in the future.

**Comment by azhiltsov** [comment]
We have faced the issue when under high load/high event rate it corrupts metrics. 
The configuration consist of 4 replicas with 24 servers within each (4x24) total. Corruption happens only on 3 of them, except first in configuration file. Metrics are getting corrupted by adding garbage at front of the metric name:
![image](https://user-images.githubusercontent.com/8709761/39129738-89d9105a-470b-11e8-9bf9-bbf1c725cdac.png)

**Comment by chbuescher** [comment]
We never used replication based on CARBON or FNV1a hashes, so we didn't see that error yet. But i've made a optimization when searching the correct position in the hash ring. Could you please test, if the errror also shows up with the unmodified consistent-hash.c from version/tag v2.6 ?

**Comment by chbuescher** [comment]
Sorry, didn't want to close this..,.
